File: HelperRequest.m
//
//  HelperRequest.m
//  Monolingual
//
//  Created by Ingmar Stein on 09.04.15.
//
//

#import "HelperRequest.h"

@implementation HelperRequest

- (instancetype)initWithCoder:(NSCoder *)coder
{
	self = [super init];
	if (self) {
		NSSet *stringArray = [NSSet setWithObjects:[NSString class], [NSArray class], nil];
		NSSet *stringSet = [NSSet setWithObjects:[NSString class], [NSSet class], nil];

		_dryRun = [coder decodeBoolForKey:@"dryRun"];
		_doStrip = [coder decodeBoolForKey:@"doStrip"];
		_uid = [coder decodeIntegerForKey:@"uid"];
		_trash = [coder decodeBoolForKey:@"trash"];
		_includes = [coder decodeObjectOfClasses:stringArray forKey:@"includes"];
		_excludes = [coder decodeObjectOfClasses:stringArray forKey:@"excludes"];
		_bundleBlacklist = [coder decodeObjectOfClasses:stringSet forKey:@"bundleBlacklist"];
		_directories = [coder decodeObjectOfClasses:stringSet forKey:@"directories"];
		_files = [coder decodeObjectOfClasses:stringArray forKey:@"files"];
		_thin = [coder decodeObjectOfClasses:stringArray forKey:@"thin"];
	}
	return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
	[coder encodeBool:self.dryRun forKey:@"dryRun"];
	[coder encodeBool:self.doStrip forKey:@"doStrip"];
	[coder encodeInteger:self.uid forKey:@"uid"];
	[coder encodeBool:self.trash forKey:@"trash"];
	[coder encodeObject:self.includes forKey:@"includes"];
	[coder encodeObject:self.excludes forKey:@"excludes"];
	[coder encodeObject:self.bundleBlacklist forKey:@"bundleBlacklist"];
	[coder encodeObject:self.directories forKey:@"directories"];
	[coder encodeObject:self.files forKey:@"files"];
	[coder encodeObject:self.thin forKey:@"thin"];
}

+ (BOOL)supportsSecureCoding {
	return YES;
}

@end

File: Helper/HelperRequest.swift
//
//  HelperRequest.swift
//  Monolingual
//
//  Created by Ingmar Stein on 10.04.15.
//
//

import Foundation

@objc(HelperRequest) class HelperRequest : NSObject, NSSecureCoding {

	var dryRun : Bool
	var doStrip : Bool
	var uid : uid_t
	var trash : Bool
	var includes : [String]?
	var excludes : [String]?
	var bundleBlacklist : Set<String>?
	var directories : Set<String>?
	var files : [String]?
	var thin : [String]?

	override init() {
		dryRun = false
		doStrip = false
		uid = 0
		trash = false

		super.init()
	}

	required init(coder aDecoder: NSCoder) {
		// https://devforums.apple.com/message/1124684
		let stringClass = NSString.self as AnyObject as! NSObject
		let arrayClass = NSArray.self as AnyObject as! NSObject
		let setClass = NSSet.self as AnyObject as! NSObject
		let stringArray = Set([stringClass, arrayClass])
		let stringSet = Set([stringClass, setClass])

		dryRun = aDecoder.decodeBoolForKey("dryRun")
		doStrip = aDecoder.decodeBoolForKey("doStrip")
		uid = uid_t(aDecoder.decodeIntegerForKey("uid"))
		trash = aDecoder.decodeBoolForKey("trash")
		includes = aDecoder.decodeObjectOfClasses(stringArray, forKey:"includes") as? [String]
		excludes = aDecoder.decodeObjectOfClasses(stringArray, forKey:"excludes") as? [String]
		bundleBlacklist = aDecoder.decodeObjectOfClasses(stringSet, forKey:"bundleBlacklist") as? Set<String>
		directories = aDecoder.decodeObjectOfClasses(stringSet, forKey:"directories") as? Set<String>
		files = aDecoder.decodeObjectOfClasses(stringArray, forKey:"files") as? [String]
		thin = aDecoder.decodeObjectOfClasses(stringArray, forKey:"thin") as? [String]

		super.init()
	}

	func encodeWithCoder(coder : NSCoder) {
		coder.encodeBool(dryRun, forKey:"dryRun")
		coder.encodeBool(doStrip, forKey:"doStrip")
		coder.encodeInteger(Int(uid), forKey:"uid")
		coder.encodeBool(trash, forKey:"trash")
		coder.encodeObject(includes, forKey:"includes")
		coder.encodeObject(excludes, forKey:"excludes")
		coder.encodeObject(bundleBlacklist, forKey:"bundleBlacklist")
		coder.encodeObject(directories, forKey:"directories")
		coder.encodeObject(files, forKey:"files")
		coder.encodeObject(thin, forKey:"thin")
	}

	static func supportsSecureCoding() -> Bool {
		return true
	}
}
